In [1]:
import pandas as pd

We will start using an insurance dataset now! This is publicly available dental insurance data from Washington State. Pandas has tons of useful ways to get data into a dataframe or series. pd.read_{format} will show you all the options; for now, we will read a CSV. .head will return the top n rows (default==5)

In [2]:
dental_data=pd.read_csv("https://data.wa.gov/api/views/gi9j-78eu/rows.csv?accessType=DOWNLOAD")
dental_data.head(5)

Company Name  NAIC Code  \
0                      4 Ever Life Insurance Company      80985   
1                     ACE AMERICAN INSURANCE COMPANY      22667   
2                       AETNA LIFE INSURANCE COMPANY      60054   
3  American Family Life Assurance Company of Colu...      60380   
4                American Fidelity Assurance Company      60410   

  Domicile State Business Type  Year  Dental Premiums  Dental Payments  \
0             IL             L  2015     1.240092e+06     1.240092e+06   
1             PA             P  2015     6.686878e+07     3.041012e+07   
2             CT             L  2015     1.621957e+09     1.243380e+09   
3             NE             L  2015     1.700630e+08     7.089238e+07   
4             OK             L  2015     1.312590e+05     3.248900e+04   

   Dental Members  Dental Member Months  Dental Loss Ratio  \
0          556973               6755924                100   
1          407257               3866500                 45   
2         3675656              44447752                 77   
3          536658               6060104                 42   
4             399                  6097                 25   

   Average Amount of Premiums per Member per Month  \
0                                             0.18   
1                                            17.29   
2                                            36.49   
3                                            28.06   
4                                            21.53   

   Previous Year Average Amount of Premiums per Member per Month  \
0                                                NaN               
1                                                NaN               
2                                                NaN               
3                                                NaN               
4                                                NaN               

   Percentage Change in Average Premium per Member per Month from Previous Year  
0                                                NaN                             
1                                                NaN                             
2                                                NaN                             
3                                                NaN                             
4                                                NaN

If you just want to select one column, slice by that column; that will return a Series, and you can use your Series methods on it

In [3]:
print(type(dental_data["Dental Premiums"]))
dental_data["Dental Premiums"].mean()

<class 'pandas.core.series.Series'>


160468754.0423341

Note that the default index on a dataframe is a zero based integer index; you can think about it like a row number. You can set the index to be something useful like the company name

In [5]:
dental_data.set_index("Company Name",inplace=True)

The recommended way to return subsets of the dataframe (which are also dataframes!) is to use .loc[row index, column index]. Use : for return all

In [6]:
dental_data.loc["AETNA LIFE INSURANCE COMPANY",:]

NAIC Code Domicile State Business Type  Year  \
Company Name                                                                 
AETNA LIFE INSURANCE COMPANY      60054             CT             L  2015   
AETNA LIFE INSURANCE COMPANY      60054             CT             L  2016   
AETNA LIFE INSURANCE COMPANY      60054             CT             L  2017   
AETNA LIFE INSURANCE COMPANY      60054             CT             L  2018   
AETNA LIFE INSURANCE COMPANY      60054             CT             L  2019   
AETNA LIFE INSURANCE COMPANY      60054             CT             L  2020   
AETNA LIFE INSURANCE COMPANY      60054             CT             L  2021   
AETNA LIFE INSURANCE COMPANY      60054             CT             L  2022   
AETNA LIFE INSURANCE COMPANY      60054             CT             L  2023   

                              Dental Premiums  Dental Payments  \
Company Name                                                     
AETNA LIFE INSURANCE COMPANY     1.621957e+09     1.243380e+09   
AETNA LIFE INSURANCE COMPANY     9.739158e+08     8.960126e+08   
AETNA LIFE INSURANCE COMPANY     1.685728e+09     1.362484e+09   
AETNA LIFE INSURANCE COMPANY     1.176694e+09     8.711933e+08   
AETNA LIFE INSURANCE COMPANY     1.198862e+09     9.500146e+08   
AETNA LIFE INSURANCE COMPANY     9.018199e+08     6.090545e+08   
AETNA LIFE INSURANCE COMPANY     9.510296e+08     7.210267e+08   
AETNA LIFE INSURANCE COMPANY     1.081450e+09     8.506224e+08   
AETNA LIFE INSURANCE COMPANY     1.146121e+09     9.223302e+08   

                              Dental Members  Dental Member Months  \
Company Name                                                         
AETNA LIFE INSURANCE COMPANY         3675656              44447752   
AETNA LIFE INSURANCE COMPANY         4252720              50996560   
AETNA LIFE INSURANCE COMPANY         4433878              53514383   
AETNA LIFE INSURANCE COMPANY         3253495              39652495   
AETNA LIFE INSURANCE COMPANY         3299239              29160856   
AETNA LIFE INSURANCE COMPANY         2564452              25763828   
AETNA LIFE INSURANCE COMPANY         1483964              15937897   
AETNA LIFE INSURANCE COMPANY         2580051              31026215   
AETNA LIFE INSURANCE COMPANY         2594887              31517841   

                              Dental Loss Ratio  \
Company Name                                      
AETNA LIFE INSURANCE COMPANY                 77   
AETNA LIFE INSURANCE COMPANY                 92   
AETNA LIFE INSURANCE COMPANY                 81   
AETNA LIFE INSURANCE COMPANY                 74   
AETNA LIFE INSURANCE COMPANY                 79   
AETNA LIFE INSURANCE COMPANY                 68   
AETNA LIFE INSURANCE COMPANY                 76   
AETNA LIFE INSURANCE COMPANY                 79   
AETNA LIFE INSURANCE COMPANY                 80   

                              Average Amount of Premiums per Member per Month  \
Company Name                                                                    
AETNA LIFE INSURANCE COMPANY                                            36.49   
AETNA LIFE INSURANCE COMPANY                                            19.10   
AETNA LIFE INSURANCE COMPANY                                            31.50   
AETNA LIFE INSURANCE COMPANY                                            29.68   
AETNA LIFE INSURANCE COMPANY                                            41.11   
AETNA LIFE INSURANCE COMPANY                                            35.00   
AETNA LIFE INSURANCE COMPANY                                            59.67   
AETNA LIFE INSURANCE COMPANY                                            34.86   
AETNA LIFE INSURANCE COMPANY                                            36.36   

                              Previous Year Average Amount of Premiums per Member per Month  \
Company Name                                                                                  
AETNA LIFE INSURANCE COMPANY 

In [7]:
dental_data.loc[["AETNA LIFE INSURANCE COMPANY","American Fidelity Assurance Company"],:]

NAIC Code Domicile State Business Type  \
Company Name                                                                  
AETNA LIFE INSURANCE COMPANY             60054             CT             L   
AETNA LIFE INSURANCE COMPANY             60054             CT             L   
AETNA LIFE INSURANCE COMPANY             60054             CT             L   
AETNA LIFE INSURANCE COMPANY             60054             CT             L   
AETNA LIFE INSURANCE COMPANY             60054             CT             L   
AETNA LIFE INSURANCE COMPANY             60054             CT             L   
AETNA LIFE INSURANCE COMPANY             60054             CT             L   
AETNA LIFE INSURANCE COMPANY             60054             CT             L   
AETNA LIFE INSURANCE COMPANY             60054             CT             L   
American Fidelity Assurance Company      60410             OK             L   
American Fidelity Assurance Company      60410             OK             L   
American Fidelity Assurance Company      60410             OK             L   
American Fidelity Assurance Company      60410             OK             L   
American Fidelity Assurance Company      60410             OK             L   

                                     Year  Dental Premiums  Dental Payments  \
Company Name                                                                  
AETNA LIFE INSURANCE COMPANY         2015     1.621957e+09     1.243380e+09   
AETNA LIFE INSURANCE COMPANY         2016     9.739158e+08     8.960126e+08   
AETNA LIFE INSURANCE COMPANY         2017     1.685728e+09     1.362484e+09   
AETNA LIFE INSURANCE COMPANY         2018     1.176694e+09     8.711933e+08   
AETNA LIFE INSURANCE COMPANY         2019     1.198862e+09     9.500146e+08   
AETNA LIFE INSURANCE COMPANY         2020     9.018199e+08     6.090545e+08   
AETNA LIFE INSURANCE COMPANY         2021     9.510296e+08     7.210267e+08   
AETNA LIFE INSURANCE COMPANY         2022     1.081450e+09     8.506224e+08   
AETNA LIFE INSURANCE COMPANY         2023     1.146121e+09     9.223302e+08   
American Fidelity Assurance Company  2015     1.312590e+05     3.248900e+04   
American Fidelity Assurance Company  2016     6.271400e+04     5.185000e+03   
American Fidelity Assurance Company  2017     3.243500e+04     6.688000e+03   
American Fidelity Assurance Company  2018     1.907100e+04     2.751000e+03   
American Fidelity Assurance Company  2019     9.580000e+02     8.490000e+02   

                                     Dental Members  Dental Member Months  \
Company Name                                                                
AETNA LIFE INSURANCE COMPANY                3675656              44447752   
AETNA LIFE INSURANCE COMPANY                4252720              50996560   
AETNA LIFE INSURANCE COMPANY                4433878              53514383   
AETNA LIFE INSURANCE COMPANY                3253495              39652495   
AETNA LIFE INSURANCE COMPANY                3299239              29160856   
AETNA LIFE INSURANCE COMPANY                2564452              25763828   
AETNA LIFE INSURANCE COMPANY                1483964              15937897   
AETNA LIFE INSURANCE COMPANY                2580051              31026215   
AETNA LIFE INSURANCE COMPANY                2594887              31517841   
American Fidelity Assurance Company             399                  6097   
American Fidelity Assurance Company             211                  3566   
American Fidelity Assurance Company             124                  1967   
American Fidelity Assurance Company               0                     0   
American Fidelity Assurance Company               0                     0   

                                     Dental Loss Ratio  \
Company Name                                             
AETNA LIFE INSURANCE COMPANY                        77   
AETNA LIFE INSURANCE COMPANY                        92   
AETNA LIFE INSURANCE COMPANY              

In [8]:
dental_data.loc["AETNA LIFE INSURANCE COMPANY","Year"]

Company Name
AETNA LIFE INSURANCE COMPANY    2015
AETNA LIFE INSURANCE COMPANY    2016
AETNA LIFE INSURANCE COMPANY    2017
AETNA LIFE INSURANCE COMPANY    2018
AETNA LIFE INSURANCE COMPANY    2019
AETNA LIFE INSURANCE COMPANY    2020
AETNA LIFE INSURANCE COMPANY    2021
AETNA LIFE INSURANCE COMPANY    2022
AETNA LIFE INSURANCE COMPANY    2023
Name: Year, dtype: int64

In [9]:
dental_data.loc[:,"Year"]

Company Name
4 Ever Life Insurance Company                                       2015
ACE AMERICAN INSURANCE COMPANY                                      2015
AETNA LIFE INSURANCE COMPANY                                        2015
American Family Life Assurance Company of Columbus (AFLAC)          2015
American Fidelity Assurance Company                                 2015
                                                                    ... 
United National Life Insurance Company of America                   2023
UNITED OF OMAHA LIFE INSURANCE COMPANY                              2023
UNITED STATES BRANCH OF THE SUN LIFE ASSURANCE COMPANY OF CANADA    2023
UnitedHealthcare Insurance Company                                  2023
Willamette Dental of Washington Inc.                                2023
Name: Year, Length: 874, dtype: int64

In [10]:
dental_data.loc[:,["Year", "Dental Premiums"]]

Year  Dental Premiums
Company Name                                                             
4 Ever Life Insurance Company                       2015     1.240092e+06
ACE AMERICAN INSURANCE COMPANY                      2015     6.686878e+07
AETNA LIFE INSURANCE COMPANY                        2015     1.621957e+09
American Family Life Assurance Company of Colum...  2015     1.700630e+08
American Fidelity Assurance Company                 2015     1.312590e+05
...                                                  ...              ...
United National Life Insurance Company of America   2023     2.303072e+06
UNITED OF OMAHA LIFE INSURANCE COMPANY              2023     1.908924e+08
UNITED STATES BRANCH OF THE SUN LIFE ASSURANCE ...  2023     2.985839e+08
UnitedHealthcare Insurance Company                  2023     1.509628e+09
Willamette Dental of Washington Inc.                2023     6.498356e+07

[874 rows x 2 columns]

To save data out, use a dataframe.to_{format} method; most useful\common are to_csv, to_excel, to_sql, to_clipboard. These all have a ton of options, and the docmentation is really good, so make sure to use it!

In [ ]:
dental_data.to_csv("dental.csv")

Dataframe.describe and .info are super useful summarizations of the data that you have

In [11]:
dental_data.describe()

NAIC Code         Year  Dental Premiums  Dental Payments  \
count    874.000000   874.000000     8.740000e+02     8.740000e+02   
mean   65363.220824  2018.872998     1.604688e+08     1.187143e+08   
std    16787.654907     2.562149     5.372048e+08     4.211788e+08   
min    11121.000000  2015.000000    -3.699800e+04    -3.709838e+06   
25%    61301.000000  2017.000000     1.273266e+06     5.405058e+05   
50%    66087.000000  2019.000000     7.367995e+06     3.886266e+06   
75%    71439.000000  2021.000000     6.068700e+07     4.047323e+07   
max    98205.000000  2023.000000     4.616818e+09     3.889293e+09   

       Dental Members  Dental Member Months  Dental Loss Ratio  \
count    8.740000e+02          8.740000e+02         874.000000   
mean     3.778362e+05          4.460064e+06          57.554920   
std      1.173658e+06          1.392633e+07          80.753268   
min      0.000000e+00          0.000000e+00       -1670.000000   
25%      2.837500e+03          3.450300e+04          44.250000   
50%      1.863200e+04          2.128815e+05          62.000000   
75%      1.567045e+05          1.857121e+06          73.000000   
max      1.072178e+07          1.225500e+08        1143.000000   

       Average Amount of Premiums per Member per Month  \
count                                       870.000000   
mean                                         47.470586   
std                                          79.671298   
min                                          -0.230000   
25%                                          27.612500   
50%                                          33.390000   
75%                                          44.107500   
max                                        1030.680000   

       Previous Year Average Amount of Premiums per Member per Month  \
count                                         760.000000               
mean                                           46.424816               
std                                            75.904441               
min                                            -0.230000               
25%                                            27.480000               
50%                                            33.350000               
75%                                            43.100000               
max                                          1025.210000               

       Percentage Change in Average Premium per Member per Month from Previous Year  
count                                         750.000000                             
mean                                            0.186813                             
std                                             1.628221                             
min                                            -2.130000                             
25%                                            -0.050000                             
50%                                             0.010000                             
75%                                             0.060000                             
max                                            30.500000

In [12]:
dental_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 874 entries, 4 Ever Life Insurance Company to Willamette Dental of Washington Inc.
Data columns (total 12 columns):
 #   Column                                                                        Non-Null Count  Dtype  
---  ------                                                                        --------------  -----  
 0   NAIC Code                                                                     874 non-null    int64  
 1   Domicile State                                                                874 non-null    object 
 2   Business Type                                                                 874 non-null    object 
 3   Year                                                                          874 non-null    int64  
 4   Dental Premiums                                                               874 non-null    float64
 5   Dental Payments                                                               874 non

From the above, we can see that `Previous Year Average Amount of Premiums per Member per Month` has a decent amount of nulls; lets drop the null rows

In [13]:
nonnull_index=dental_data["Previous Year Average Amount of Premiums per Member per Month"].notnull()
nonnull_index

Company Name
4 Ever Life Insurance Company                                       False
ACE AMERICAN INSURANCE COMPANY                                      False
AETNA LIFE INSURANCE COMPANY                                        False
American Family Life Assurance Company of Columbus (AFLAC)          False
American Fidelity Assurance Company                                 False
                                                                    ...  
United National Life Insurance Company of America                    True
UNITED OF OMAHA LIFE INSURANCE COMPANY                               True
UNITED STATES BRANCH OF THE SUN LIFE ASSURANCE COMPANY OF CANADA     True
UnitedHealthcare Insurance Company                                   True
Willamette Dental of Washington Inc.                                 True
Name: Previous Year Average Amount of Premiums per Member per Month, Length: 874, dtype: bool

When you pass a boolean series (i.e. series of True's and False's to .loc, it will only keep the rows where the value is true)

In [14]:
dental_data.loc[nonnull_index,:]

NAIC Code Domicile State  \
Company Name                                                                   
4 Ever Life Insurance Company                           80985             IL   
ACE AMERICAN INSURANCE COMPANY                          22667             PA   
AETNA LIFE INSURANCE COMPANY                            60054             CT   
American Family Life Assurance Company of Colum...      60380             NE   
American Fidelity Assurance Company                     60410             OK   
...                                                       ...            ...   
United National Life Insurance Company of America       92703             NE   
UNITED OF OMAHA LIFE INSURANCE COMPANY                  69868             NE   
UNITED STATES BRANCH OF THE SUN LIFE ASSURANCE ...      80802             MI   
UnitedHealthcare Insurance Company                      79413             CT   
Willamette Dental of Washington Inc.                    47050             WA   

                                                   Business Type  Year  \
Company Name                                                             
4 Ever Life Insurance Company                                  L  2016   
ACE AMERICAN INSURANCE COMPANY                                 P  2016   
AETNA LIFE INSURANCE COMPANY                                   L  2016   
American Family Life Assurance Company of Colum...             L  2016   
American Fidelity Assurance Company                            L  2016   
...                                                          ...   ...   
United National Life Insurance Company of America              L  2023   
UNITED OF OMAHA LIFE INSURANCE COMPANY                         L  2023   
UNITED STATES BRANCH OF THE SUN LIFE ASSURANCE ...             L  2023   
UnitedHealthcare Insurance Company                             L  2023   
Willamette Dental of Washington Inc.                           X  2023   

                                                    Dental Premiums  \
Company Name                                                          
4 Ever Life Insurance Company                          1.363989e+06   
ACE AMERICAN INSURANCE COMPANY                         6.798916e+07   
AETNA LIFE INSURANCE COMPANY                           9.739158e+08   
American Family Life Assurance Company of Colum...     1.665599e+08   
American Fidelity Assurance Company                    6.271400e+04   
...                                                             ...   
United National Life Insurance Company of America      2.303072e+06   
UNITED OF OMAHA LIFE INSURANCE COMPANY                 1.908924e+08   
UNITED STATES BRANCH OF THE SUN LIFE ASSURANCE ...     2.985839e+08   
UnitedHealthcare Insurance Company                     1.509628e+09   
Willamette Dental of Washington Inc.                   6.498356e+07   

                                                    Dental Payments  \
Company Name                                                          
4 Ever Life Insurance Company                          1.363989e+06   
ACE AMERICAN INSURANCE COMPANY                         4.089506e+07   
AETNA LIFE INSURANCE COMPANY                           8.960126e+08   
American Family Life Assurance Company of Colum...     7.147889e+07   
American Fidelity Assurance Company                    5.185000e+03   
...                                                             ...   
United National Life Insurance Company of America      1.053156e+06   
UNITED OF OMAHA LIFE INSURANCE COMPANY                 1.399700e+08   
UNITED STATES BRANCH OF THE SUN LIFE ASSURANCE ...     2.107214e+08   
UnitedHealthcare Insurance Company                     1.157232e+09   
Willamette Dental of Washington Inc.                   5.907501e+07   

                                                    Dental Members  \
Company Name                                                         
4 Ever Life Insurance Company                               

760 is the right number; you can also use len(dataframe) to get the rowcount, or dataframe.shape to get row and column count

In [ ]:
dental_data.shape